js/xeval.ts

```diff
@@ -0,0 +1,76 @@
+import { Buffer } from "./buffer";
+import { stdin } from "./files";
+import { TextEncoder, TextDecoder } from "./text_encoding";
+import { exit } from "./os";
+
+export type XevalFunc = (v: string) => void;
+
+async function writeAll(buffer: Buffer, arr: Uint8Array): Promise<void> {
+  let bytesWritten = 0;
+  while (bytesWritten < arr.length) {
+    try {
+      const nwritten = await buffer.write(arr.subarray(bytesWritten));
+      bytesWritten += nwritten;
+    } catch {
+      return;
+    }
+  }
+}
+
+export async function xevalMain(
+  xevalFunc: XevalFunc,
+  delim_: string | null
+): Promise<void> {
+  const inputBuffer = new Buffer();
+  const inspectArr = new Uint8Array(1024);
+  const encoder = new TextEncoder();
+  const decoder = new TextDecoder();
+  if (!delim_) {
+    delim_ = "\n";
+  }
+  const delimArr = encoder.encode(delim_); // due to unicode
+
+  let nextMatchIndex = 0;
+  while (true) {
+    const rr = await stdin.read(inspectArr);
+    if (rr.nread < 0) {
+      exit(1);
+    }
+    const sliceRead = inspectArr.subarray(0, rr.nread);
+    // Avoid unicode problems
+    let nextSliceStartIndex = 0;
+    for (let i = 0; i < sliceRead.length; i++) {
```

Comment:
Can you abstract out a function called `lines()` which yields a generator of strings. I believe @bartlomieju might have done a version of this before. Add a TODO above function stating that this functionality exists in deno_std, and that we should dedup the code once we have the ability to import from there.

Reply:
Did not find this utility in `deno_std`. I'll add a comment nevertheless, saying we might want to port this to `deno_std` and later import from it once it is doable.

Reply:
I made a try in #2101 but we didn't land it.

Reply:
https://github.com/denoland/deno_std/blob/aa9446390230da43d5ac6dcdcf8c355e1deedb12/textproto/mod.ts#L39-L45

Reply:
@ry I instead abstracted out `chunks(reader, delim)` that supports any string delimiter of any length.